src/pages/VolunteerPage.jsx

```diff
@@ -0,0 +1,288 @@
+    const [filteredShelters, setFilteredShelters] = useState([]);
+    const [loading, setLoading] = useState(false);
+    const [error, setError] = useState(null);
+    const [selectedDate, setSelectedDate] = useState("");
+    const [endDate, setEndDate] = useState(""); // New state for end date
+    const [selectedHostId, setSelectedHostId] = useState("");
+    const [hosts, setHosts] = useState([]);
+    const [showPopover, setShowPopover] = useState(false);
+    const [selectedProduct, setSelectedProduct] = useState(null);
+    const [userFirstName, setUserFirstName] = useState("");
+    const [userLastName, setUserLastName] = useState("");
+    const [foundUserId, setFoundUserId] = useState(null);
+    const [foundUser, setFoundUser] = useState(null);
+    const [searchError, setSearchError] = useState(null);
+
+    useEffect(() => {
+        const fetchAllShelters = async () => {
+            setLoading(true);
+            try {
+                const response = await axios.get('/api/volunteer/available');
+                const shelters = response.data;
+                setAvailableShelters(shelters);
+
+                const uniqueHosts = [];
+                const hostIds = new Set();
+                shelters.forEach(shelter => {
+                    if (!hostIds.has(shelter.host.id)) {
+                        uniqueHosts.push(shelter.host);
+                        hostIds.add(shelter.host.id);
+                    }
+                });
+                setHosts(uniqueHosts);
+            } catch (err) {
+                setError('Failed to load available shelters');
+            } finally {
+                setLoading(false);
+            }
+        };
+        fetchAllShelters();
+    }, []);
+
+    const handleFilter = () => {
+        let filtered = [...availableShelters];
+        if (selectedHostId) {
+            filtered = filtered.filter(shelter => shelter.host.id === parseInt(selectedHostId));
+        }
+        if (selectedDate) {
+            filtered = filtered.map(shelter => ({
+                ...shelter,
+                products: shelter.products.filter(product => {
+                    const availability = product.availability?.find(avail => avail.date === selectedDate);
+                    product.places_left = availability ? availability.places_left : product.total_places;
+                    return product.places_left > 0;
+                })
+            })).filter(shelter => shelter.products.length > 0);
+        }
+        setFilteredShelters(filtered);
+    };
+
+    const openBookingPopover = (product) => {
+        setSelectedProduct(product);
+        setShowPopover(true);
+    };
+
+    const closePopover = () => {
+        setShowPopover(false);
+        setSelectedProduct(null);
+        setUserFirstName("");
+        setUserLastName("");
+        setFoundUser(null);
+        setFoundUserId(null);
+        setSearchError(null);
+    };
+
+    const handleBooking = async () => {
```

Comment:
After booking is confirmed, update the `avalibleShelters` state to reflect the reservation.
setting `places_left` to 0 or removing it if fully booked, ensuring the UI reflects the most recent availability.